In [8]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [9]:
path= r"c:\users\harip\Instacart basket analysis"

In [10]:
df_prods= pd.read_csv(os.path.join(path,'02 Data','Original Data', 'products.csv'), index_col = False)

In [11]:
df_ords= pd.read_csv(os.path.join(path,'02 data','prepared data','orders_wrangled.csv'), index_col = 0)

# Consistency checks for df_prods

In [12]:
#checking for mixed data
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [13]:
#finding the missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prods.shape

(49693, 5)

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
#Finding the duplicate data
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum value of prices column appears to be higher and needs a special focus.

In [26]:
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

There might be a mistake as a 2% reduced fat milk can't be in the price range.

In [27]:
# Check for products that cost more than $50
df_prods_clean_no_dups[df_prods_clean_no_dups["prices"]> 50.0]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

Its so obvious that this two products may have an error in decimal points.

In [28]:
#replacing the two values in the prices section
df_prods_clean_no_dups = df_prods_clean_no_dups.replace ({"prices":{99999.0: 9.99, 14900.0:1.49 }})

In [29]:
df_prods_clean_no_dups[df_prods_clean_no_dups['product_id'] == 33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    9.99

In [30]:
df_prods_clean_no_dups[df_prods_clean_no_dups['product_id'] == 21553]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

       prices  
21554    1.49

The changes have been made in the product_id 33664 and 21553.

# Consistency checks on df_ords

In [31]:
#determing the values
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       orders_timing_in_day  days_since_prior_order  
count          3.421083e+06            3.214874e+06  
mean           1.345202e+01            1.111484e+01  
std            4.226088e+00            9.206737e+00  
min            0.000000e+00            0.000000e+00  
25%            1.000000e+01            4.000000e+00  
50%            1.300000e+01            7.000000e+00  
75%            1.600000e+01            1.500000e+01  
max            2.300000e+01            3.000000e+01

The descriptive statistics appears to be normal.

# Checking for Mixed data types

In [32]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Observations: No mixed data detected

# Finding Missing values

In [33]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
orders_timing_in_day           0
days_since_prior_order    206209
dtype: int64

In [34]:
#creating a subset for missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [35]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         orders_timing_in_day  days_since_prior_order  
0                           8                     NaN  
11                         11                     NaN  
26                         14                     NaN  
39                         11                     NaN  
45                         12                     NaN  
...                       ...                     ...  
3420930                    12                     NaN  
3420934                    18                     NaN  
3421002                    19                     NaN  
3421019                    15                     NaN  
3421069                    11                     NaN  

[206209 rows x 6 columns]

In [36]:
#Addressing the missing values

In my opinion, I chose to continue without imputing or removing the missing values since it has almost 206209 rows missing so by imputing or removing the data there will be massive loss in data will end in improper analysis.

# Finding duplicate data

In [37]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, orders_timing_in_day, days_since_prior_order]
Index: []

No duplicates seen

In [39]:
df_ords.shape

(3421083, 6)

# Exporting data

In [40]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [42]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))